libcst/codemod/visitors/_gather_string_annotation_names.py

```diff
@@ -35,14 +35,48 @@ class GatherNamesFromStringAnnotationsVisitor(ContextAwareVisitor):
 
     METADATA_DEPENDENCIES = (QualifiedNameProvider,)
 
-    def __init__(self, context: CodemodContext) -> None:
+    def __init__(
+        self, context: CodemodContext, typing_functions: Optional[Set[str]] = None
+    ) -> None:
         super().__init__(context)
 
+        self._typing_functions: Set[str] = (
+            typing_functions
+            if typing_functions is not None
+            else FUNCS_CONSIDERED_AS_STRING_ANNOTATIONS
+        )
+        self._annotation_stack: List[cst.CSTNode] = []
         #: The set of names collected from string literals.
         self.names: Set[str] = set()
 
-    @m.call_if_inside(ANNOTATION_MATCHER)
-    @m.visit(m.ConcatenatedString())
+    def visit_Annotation(self, node: cst.Annotation) -> bool:
+        self._annotation_stack.append(node)
+        return True
+
+    def leave_Annotation(self, original_node: cst.Annotation) -> None:
+        self._annotation_stack.pop()
+
+    def visit_Call(self, node: cst.Call) -> bool:
+        qnames = self.get_metadata(QualifiedNameProvider, node)
+        if any(qn.name in self._typing_functions for qn in qnames):
+            self._annotation_stack.append(node)
+            return True
+        return False
+
+    def leave_Call(self, original_node: cst.Call) -> None:
+        if self._annotation_stack != [] and self._annotation_stack[-1] == original_node:
+            self._annotation_stack.pop()
+
+    def visit_ConcatenatedString(self, node: cst.ConcatenatedString) -> bool:
+        if self._annotation_stack != []:
+            self.handle_any_string(node)
+        return False
+
+    def visit_SimpleString(self, node: cst.SimpleString) -> bool:
+        if self._annotation_stack != []:
+            self.handle_any_string(node)
+        return False
```

Comment:
Why not simply check the stack as as boolean:

```suggestion
    def leave_Call(self, original_node: cst.Call) -> None:
        if self._annotation_stack and self._annotation_stack[-1] == original_node:
            self._annotation_stack.pop()

    def visit_ConcatenatedString(self, node: cst.ConcatenatedString) -> bool:
        if self._annotation_stack:
            self.handle_any_string(node)
        return False

    def visit_SimpleString(self, node: cst.SimpleString) -> bool:
        if self._annotation_stack:
            self.handle_any_string(node)
        return False
```